Psi Alert

walksanatora opened the pull request:
## Short description
Adds a new alert level JUST for dealing with silicon-based threats

## Why we need to add this
I forsee it becoming a issue dealing with xenoborgs, malf AI, a particularly robust derelict borg. this alert level sends a new Psi code armory filled with EMP and anti-silicon turrets just to destroy tresspassers. also contains a refactor for how sending gamma armory is handeled and now allows defining custom armories + announcements for other alert levels. (Omega armory anyone?)

## Media (Video/Screenshots)
<img width="402" height="186" alt="image" src="https://github.com/user-attachments/assets/b35dcf24-e9b7-4f01-bdcf-34165b679258" />
<img width="270" height="114" alt="image" src="https://github.com/user-attachments/assets/dc0d665e-a712-402a-8dee-6df5733578f8" />
<img width="398" height="423" alt="image" src="https://github.com/user-attachments/assets/5807a638-cb01-4cea-8506-613516e8b404" />
ignore the ERROR's I dont have a sprite for the PSI alert level on screen yet.

## Checks

- [x] I do not require assistance to complete the PR.
- [x] Before posting/requesting review of a PR, I have verified that the changes work.
- [x] I have added screenshots/videos of the changes, or this PR does not change in-game mechanics.
- [x] I affirm that my changes are licensed under the [Starlight Fork License](https://github.com/ss14Starlight/space-station-14/blob/Starlight/LICENSE-Starlight.TXT) and grant permission for use in this repository under its conditions.

**Changelog**
:cl: STARLIGHT TEAM
- add: Tired of the black and blue xenoborgs wiping out your station? just call for Psi alert and deal with it yourself.
- tweak: Alert Level armories now return to CentComm when the threat is over. kicking people off the pod.

walksanatora committed:
psi code armory

walksanatora commented:
updated the psi armory with some anti-xenoborg armor. 
<img width="377" height="412" alt="image" src="https://github.com/user-attachments/assets/17199a03-4b72-47b4-98cd-40602b25b15c" />
<img width="389" height="179" alt="image" src="https://github.com/user-attachments/assets/6925e99f-b50e-4ca2-9323-308fb3be4325" />
<img width="280" height="83" alt="image" src="https://github.com/user-attachments/assets/7c0858f6-9ee5-4e35-b9c2-26fe09ab6426" />

walksanatora committed:
reflective stuffs

Littlemen07 commented:
I am all for specific alert/paperwork/gear for specific threats, however i do not know the balancing on xenoborg (or when they will be added too)

Arent Xenoborg only able to use laser? If so having a suit with 100% reflect is just giving total immunity, its very unfair to them.

VileSloth commented:
Is adding another alert armory really needed? the gamma alert for nukies makes sense because sometimes they take the armory, but adding a special armory for another antag i feel like it may trivialize the antag? i hope im getting across what i mean.

walksanatora commented:
> Is adding another alert armory really needed? the gamma alert for nukies makes sense because sometimes they take the armory, but adding a special armory for another antag i feel like it may trivialize the antag? i hope im getting across what i mean.

have you seen *either* of the replays? the current gamma armory just gets stolen cause is has zero protection. they just jaws of life and and throw everything into space.

walksanatora commented:
> Arent Xenoborg only able to use laser? If so having a suit with 100% reflect is just giving total immunity, its very unfair to them.

they can only use laser for ranged. but honestly I think reflect is the LEAST of the worries in that armory. (see: the 24 grenades shootable from the hydra, the 4 ion rifles, and the I wanna say 6 emp grenades?

Rinary1 commented on Content.Server/_Starlight/AlertArmory/AlertArmoryStationComponent.cs:
```diff
@@ -0,0 +1,32 @@
+[DataDefinition]
+public sealed partial class AlertArmoryDefinition
+{
+    [DataField(required: true)]
+    public ResPath Shuttle;
+
+    [DataField]
+    public string? Announcement;
+
+    [DataField]
+    public Color? AnnouncementColor;
```
```suggestion
    ///<summary>
    /// Path to shuttle grid which will be used for spawn.
    ///</summary>
    [DataField(required: true)]
    public ResPath Shuttle;

    ///<summary>
    /// Announcement message which will be sended when armory fly to station.
    /// </summary>
    [DataField]
    public string? Announcement;

    ///<summary>
    /// Color of announcement message which will be sended when armory fly to station.
    ///</summary>
    [DataField]
    public Color? AnnouncementColor;
```

Rinary1 commented on Content.Server/_Starlight/AlertArmory/AlertArmorySystem.cs:
```diff
@@ -0,0 +1,128 @@
+                colorOverride: comp.AnnouncementColor ?? Color.PaleVioletRed);
+    }
+
+    private void OnAlertLevelChanged(AlertLevelChangedEvent ev)
```
```suggestion
    ///<summary>
    /// Handles alert level changing to try ftl dock gamma alert.
    ///</summary>
    private void OnAlertLevelChanged(AlertLevelChangedEvent ev)
```

Rinary1 commented on Content.Server/_Starlight/AlertArmory/AlertArmorySystem.cs:
```diff
@@ -0,0 +1,128 @@
+        ev.Tag = "DockGamma";
+    }
+    
+    private void AnnounceShuttleDocking(EntityUid uid, AlertArmoryShuttleComponent comp, ref FTLCompletedEvent ev)
```
```suggestion
    ///<summary>
    /// Sends announcement message about shuttle docking.
    ///</summary>
    private void AnnounceShuttleDocking(EntityUid uid, AlertArmoryShuttleComponent comp, ref FTLCompletedEvent ev)
```

Rinary1 commented on Content.Server/_Starlight/AlertArmory/AlertArmorySystem.cs:
```diff
@@ -0,0 +1,128 @@
+        }
+    }
+    
+    private void OnStartup(EntityUid uid, AlertArmoryShuttleComponent comp, ComponentStartup ev) => EnsureComp<PreventPilotComponent>(uid);
```
```suggestion
    ///<summary>
    /// Adds component so you can't pilot alert armory and fly off on it
    ///</summary>
    private void OnStartup(EntityUid uid, AlertArmoryShuttleComponent comp, ComponentStartup ev) => EnsureComp<PreventPilotComponent>(uid);
```

Rinary1 commented on Content.Server/_Starlight/AlertArmory/AlertArmorySystem.cs:
```diff
@@ -0,0 +1,128 @@
+        SubscribeLocalEvent<AlertLevelChangedEvent>(OnAlertLevelChanged);
+    }
+    
+    private void InitializeGammaWeaponryStation(EntityUid uid, AlertArmoryStationComponent comp, StationPostInitEvent ev)
```
```suggestion
    ///<summary>
    /// Initialize station with armories and preload all armories.
    ///</summary>
    private void InitializeAlertArmoryStation(EntityUid uid, AlertArmoryStationComponent comp, StationPostInitEvent ev)
```

Rinary1 commented on Content.Server/_Starlight/AlertArmory/AlertArmorySystem.cs:
```diff
@@ -0,0 +1,128 @@
+    
+    public override void Initialize()
+    {
+        SubscribeLocalEvent<AlertArmoryStationComponent, StationPostInitEvent>(InitializeGammaWeaponryStation);
```
```suggestion
        SubscribeLocalEvent<AlertArmoryStationComponent, StationPostInitEvent>(InitializeAlertArmoryStation);
```

Rinary1 commented on Content.Server/_Starlight/AlertArmory/AlertArmorySystem.cs:
```diff
@@ -0,0 +1,128 @@
+    private void OnStartup(EntityUid uid, AlertArmoryShuttleComponent comp, ComponentStartup ev) => EnsureComp<PreventPilotComponent>(uid);
+    
+
+    private void SetShuttleTagToDockGamma(EntityUid uid, AlertArmoryShuttleComponent comp, ref FTLTagEvent ev)
+    {
+        if (ev.Handled)
+            return;
+
+        ev.Handled = true;
+        ev.Tag = "DockGamma";
+    }
```
```suggestion
    ///<summary>
    /// Sets shuttle dock tag, so it try dock to correct place
    ///</summary>
    private void SetShuttleTag(EntityUid uid, AlertArmoryShuttleComponent comp, ref FTLTagEvent ev)
    {
        if (ev.Handled || comp.DockTag == null)
            return;

        ev.Handled = true;
        ev.Tag = comp.DockTag;
    }
```

Rinary1 commented on Content.Server/_Starlight/AlertArmory/AlertArmorySystem.cs:
```diff
@@ -0,0 +1,128 @@
+    {
+        SubscribeLocalEvent<AlertArmoryStationComponent, StationPostInitEvent>(InitializeGammaWeaponryStation);
+        SubscribeLocalEvent<AlertArmoryShuttleComponent, ComponentStartup>(OnStartup);
+        SubscribeLocalEvent<AlertArmoryShuttleComponent, FTLTagEvent>(SetShuttleTagToDockGamma);
```
```suggestion
        SubscribeLocalEvent<AlertArmoryShuttleComponent, FTLTagEvent>(SetShuttleTag);
```

walksanatora committed:
Update Content.Server/_Starlight/AlertArmory/AlertArmoryStationComponent.cs

Co-authored-by: Rinary <rinary.super@gmail.com>

walksanatora committed:
Apply suggestions from code review

Co-authored-by: Rinary <rinary.super@gmail.com>

walksanatora committed:
a little more glue for making gamma leave when it is no longer needed.

Rinary1 commented on Content.Server/_Starlight/AlertArmory/AlertArmorySystem.cs:
```diff
@@ -0,0 +1,185 @@
+        if (!set.Contains(ev.AlertLevel))
+            return;
+
+        var shuttle = comp.Grids[ev.AlertLevel];
```
```suggestion
        var shuttle = comp.Grids[ev.AlertLevel];
        
        if (!TryComp<AlertArmoryShuttleComponent>(shuttle, out var shuttleComp)
            return;
        
```

walksanatora replied:
I dont try comp because by THIS point all the grids were made by the Init event. so I KNOW they have them. and if a admeme touches this and screws it up. it is on *them* not me :>

Rinary1 replied:
> I dont try comp because by THIS point all the grids were made by the Init event. so I KNOW they have them. and if a admeme touches this and screws it up. it is on _them_ not me :>

I added try comp because it used for taking DockTag lol, not because I want to check if admeme are bad(For this we can use HasComp lol)

Rinary1 commented on Content.Server/_Starlight/AlertArmory/AlertArmorySystem.cs:
```diff
@@ -0,0 +1,185 @@
+            shuttle,
+            Comp<ShuttleComponent>(shuttle),
+            targetGrid.Value,
+            priorityTag: _gammaDock);
```
```suggestion
            priorityTag: shuttleComp.DockTag);
```

Rinary1 commented on Content.Server/_Starlight/AlertArmory/AlertArmorySystem.cs:
```diff
@@ -0,0 +1,185 @@
+    private EntityQuery<ArrivalsBlacklistComponent> _blacklistQuery;
+    private EntityQuery<MobStateComponent> _mobQuery;
+
+    private readonly ProtoId<TagPrototype> _gammaDock = "DockGamma";
```
```suggestion
```

moved to component

walksanatora committed:
armories now kick you off and return to "Centcomm" for restocking.

walksanatora committed:
Update Content.Server/_Starlight/AlertArmory/AlertArmorySystem.cs

Co-authored-by: Rinary <rinary.super@gmail.com>

walksanatora committed:
Update Content.Server/_Starlight/AlertArmory/AlertArmorySystem.cs

Co-authored-by: Rinary <rinary.super@gmail.com>

walksanatora committed:
use ProtoId

walksanatora committed:
hit it with the formatter. will this save it?

walksanatora committed:
abuse of Comp<>

Rinary1 commented on Content.Server/AlertLevel/AlertLevelSystem.cs:
```diff
@@ -206,7 +207,7 @@ public void SetLevel(EntityUid station, string level, bool playSound, bool annou
         }
 
-        RaiseLocalEvent(new AlertLevelChangedEvent(station, level));
+        RaiseLocalEvent(new AlertLevelChangedEvent(station, level, oldLevel));
```
```suggestion
        RaiseLocalEvent(new AlertLevelChangedEvent(station, level, oldLevel)); // Starlight-edit: Add Old Level
```

Rinary1 commented on Content.Server/AlertLevel/AlertLevelSystem.cs:
```diff
@@ -220,10 +221,12 @@ public sealed class AlertLevelChangedEvent : EntityEventArgs
+    public string OldAlertLevel { get;  } //Starlight
 
-    public AlertLevelChangedEvent(EntityUid station, string alertLevel)
+    public AlertLevelChangedEvent(EntityUid station, string alertLevel, string oldAlertLevel)
```
```suggestion
    public AlertLevelChangedEvent(EntityUid station, string alertLevel, string oldAlertLevel) // Starlight-edit: Add Old Level
```

Rinary1 commented on Content.Server/AlertLevel/AlertLevelSystem.cs:
```diff
@@ -220,10 +221,12 @@ public sealed class AlertLevelChangedEvent : EntityEventArgs
     {
         Station = station;
         AlertLevel = alertLevel;
+        OldAlertLevel = oldAlertLevel;
```
```suggestion
        OldAlertLevel = oldAlertLevel; // Starlight-edit: Add Old Level
```

Rinary1 commented on Content.Server/Shuttles/Systems/ArrivalsSystem.cs:
```diff
@@ -426,10 +426,8 @@ private bool TryTeleportToMapSpawn(EntityUid player, EntityUid stationId, Transf
-    {
-        EnsureComp<PreventPilotComponent>(uid);
-    }
+    private void OnShuttleStartup(EntityUid uid, ArrivalsShuttleComponent component, ComponentStartup args) => EnsureComp<PreventPilotComponent>(uid);
```
```suggestion
    private void OnShuttleStartup(EntityUid uid, ArrivalsShuttleComponent component, ComponentStartup args) => EnsureComp<PreventPilotComponent>(uid); // Starlight-edit: to lambda
```

walksanatora replied:
oops thought this was one of our functions before I read the file. reverted.

walksanatora committed:
Apply suggestions from code review

Co-authored-by: Rinary <rinary.super@gmail.com>

walksanatora committed:
revert that.

walksanatora committed:
new psi armory

walksanatora committed:
fix typo in alert level.

walksanatora committed:
Psi Armory is Opaque and cannot be shot into.

walksanatora commented:
https://github.com/user-attachments/assets/479c17ed-2dbd-49fb-9d10-d7efbfb284ee

walksanatora commented:
since that video I fixed the type in th announcement and removed the windows of the gamma armory blocking the turret. the armory is now fully opaque and cannot be seen into.

walksanatora committed:
nerf reflect suit chances so it isn't a hard counter.

walksanatora committed:
Update Psi Armory.

walksanatora commented:
<img width="355" height="377" alt="image" src="https://github.com/user-attachments/assets/7fdbdba3-4dc1-4c4a-b5c0-62e34b4316dd" />

walksanatora committed:
update armory, add intellicard, fix trims

walksanatora committed:
fix yaml linter?

rustdevelopment commented:
I think the alert color should be something other than blue, since we already have a blue alert. Maybe you could try it with orange, since the psi armory has blue and orange colors already.

walksanatora commented:
we kinda allready have a orange/gold alert. omega. ofc you wont *see* omega often cause omega is universe-ending event and is basically unusable.

walksanatora commented:
I just like the blue because it is *xenoborg blue* basically the alert level is antag colored.

stickycheZ101 commented:
> we kinda allready have a orange/gold alert. omega. ofc you wont _see_ omega often cause omega is universe-ending event and is basically unusable.

Omega is black/dark grey

walksanatora commented:
I was mainly referring to the alert screen color. (as thats the color I care the most about tbh)

walksanatora commented:
<img width="128" height="109" alt="image" src="https://github.com/user-attachments/assets/6f8dfd45-77c7-4592-89d2-7d50b2e7e0b9" />

stickycheZ101 commented:
Oh yeah true (i just thought black and gold would be a really nice scheme)

Murrytmds commented:
> we kinda allready have a orange/gold alert. omega. ofc you wont _see_ omega often cause omega is universe-ending event and is basically unusable.

honestly do the colors even matter for non-color related alerts? or even for some of them?
Hell our violet alert isn't even violet.

Conflee commented:
> honestly do the colors even matter for non-color related alerts? or even for some of them? Hell our violet alert isn't even violet.

Violet is a brighter purple because alerts need to be readable in chat mostly.

walksanatora commented:
for me I like the color of the alert to be tied to the alert in some way. Omega is black. the end of all things. Psi is Blue cause it is specifically for xenoborg threats. Gamma may be like a salmon but I can logic that as a "flesh" color.

walksanatora commented:
> Violet is a brighter purple because alerts need to be readable in chat mostly.

they are probally talking about the flashing red biohazard symbol.

rustdevelopment commented:
> > we kinda allready have a orange/gold alert. omega. ofc you wont _see_ omega often cause omega is universe-ending event and is basically unusable.
> 
> honestly do the colors even matter for non-color related alerts? or even for some of them? Hell our violet alert isn't even violet.

I just think it would be cool & novel to have some kind of new alert color, if we're adding a new type of alert
